src/com/griddynamics/devops/mpl/MPLConfig.groovy

```diff
@@ -0,0 +1,176 @@
+   * CFG is just a config place, you can't iterate over it.
+   * The module should know the key it's trying to use, not to process everything.
+   */
+  public Object remove(Object key) {}
+  public Collection values() {}
+  public Set keySet() {}
+  public void clear() {}
+  public boolean containsKey(Object key) {}
+  public boolean containsValue(Object val) {}
+  public void putAll(Map map) {}
+
+  public int size() { this.@config.size() }
+  public boolean isEmpty() { this.@config.isEmpty() }
+  public Set entrySet() { Helper.configEntrySet(this.@config) }
+  
+  /**
+   * Get a value copy of the provided config key path
+   *
+   * Warning: it's not a good idea to use the method directly as
+   *          `CFG.get('key')` - please use `CFG.'key'` instead.
+   *          In the future versions direct access could be removed
+   *
+   * Allowed path tokens:
+   *  string - almost any charset not containing dot, key of the Map
+   *  number - index of a List, if target is not Map
+   *
+   * @param key_path - path to the desired config value separated by dot
+   * @return value of the key path or null if not found
+   */
+  public Object get(Object key_path) {
+    def key_list = key_path.tokenize(".")
+
+    def value = this.@config
+    for( def key in key_list ) {
+      if( value in Map )
+        value = value[key]
+      else if( value in List && key.isInteger() && key.toInteger() >= 0 )
+        value = value[key.toInteger()]
+      else
+        value = null
+
+      if( value == null )
+        break
+    }
+    return Helper.cloneValue(value)
+  }
+
+  /**
+   * Set value for the provided config key path
+   *
+   * Warning: it's not a good idea to use the method directly as
+   *          `CFG.put('key', 'value')` - please use `CFG.'key' = 'value'` instead
+   *          In the future versions direct access could be removed
+   *
+   * Allowed path tokens:
+   *  string - almost any charset not containing dot, key of the Map
+   *  number - index of a List, if target is not Map
+   *
+   * Not existing key will be created as Map (even if number key is provided)
+   * Exception could be thrown if:
+   *  - you trying to set sublevel of the existing key which is not the required type:
+   *    + List requires positive integer: string keys will cause exception
+   *    + Non-container variables (of course)
+   *
+   * @param key_path - path to the desired config value separated by dot
+   */
+  public Object put(Object key_path, Object val) {
+    def key_list = key_path.tokenize(".")
+
+    def key, parent
+    def value = this.@config
+    for( def i = 0; i < key_list.size(); i++ ) {
+      key = key_list[i]
+      parent = value
+
+      if( parent in List ) {
+        if( key.isInteger() && key.toInteger() >= 0 )
+          key = key.toInteger()
+        else
+          throw new MPLException("Invalid config key path '${key_list[i-1] = '<'+key_list[i-1]+'>'; key_list.join('.')}': List key value is not a positive integer")
```

Comment:
Yeah this one is tricky. Check it out:
```
cfg = [a: [b: [1,2,3,[g:5]]]]
key_list = ['a', 'b', 'L', 'c']
# Exception on i = 2, because it's not a positive number
key_list[i-1] = '<'+key_list[i-1]+'>'
# Will result in key_list == ['a', '<b>', 'L', 'c']
```
We marking a key that is actually represents a list - not the key caused the issue. Probably it's not the right logic... Need to check.

Reply:
Probably it's just a copypaste from the `is not suitable to set the nested variable` exception where we actually need to mark the previous item. So let me fix this one.

Reply:
Fixed